add azure role definition module

yungezz opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
new module for azure role definition. Test requires test service principal has `owner` role in test subscription.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

yungezz committed:
add role definition module

yungezz committed:
fix sample

yungezz commented:
role definition creating failed because of test service principal doesn't have permission

https://app.shippable.com/github/ansible/ansible/runs/108108/82/console

```
"msg": "Error creating role definition: Azure Error: AuthorizationFailed\nMessage: The client '45641ef5-4510-4c63-897e-5a9ddba13dab' with object id '45641ef5-4510-4c63-897e-5a9ddba13dab' does not have authorization to perform action 'Microsoft.Authorization/roleDefinitions/write' over scope '/subscriptions/6d22db98-3e5f-4ab9-bdf9-2f911a2775f7/resourceGroups/ansible-core-ci-prod-7b90d1d7-ea55-4ef4-9769-c5fe3e25c112-1/providers/Microsoft.Authorization/roleDefinitions/14532484-50c6-40fe-949e-96db2cd028da'."
```

HI @mattclay  could you pls grant `owner` permissions? any concern?

yungezz committed:
fix lint

yungezz committed:
fix lint

yungezz committed:
add facts module

yungezz committed:
fix lint

mattclay commented:
@yungezz I'm not seeing that authorization issue on the latest test runs. Is it still a problem?

yungezz committed:
disable test due to no owner permission

mattclay commented on test/integration/targets/azure_rm_roledefinition/aliases:
```diff
@@ -1,3 +1,4 @@
 cloud/azure
 shippable/azure/group1
 destructive
+disabled
```
Use `unsupported` instead of `disabled`. The use of `disabled` is for tests which are temporarily broken, while `unsupported` is for tests we cannot run in CI.

yungezz replied:
fixed

yungezz committed:
use unsupported

yungezz commented:
add manual test output here for reference:

playbook code snippet:
```
   - name: Create a role definition (Check Mode)
      azure_rm_roledefinition:
        name: "{{ role_name }}"
        scope: "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
        permissions:
          - actions:
              - "Microsoft.Compute/virtualMachines/read"
            data_actions:
              - "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
        assignable_scopes:
          - "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
      check_mode: yes
      register: output

    - name: Assert creating role definition check mode
      assert:
        that:
          - output.changed

    - name: Create a role definition
      azure_rm_roledefinition:
        name: "{{ role_name }}"
        scope: "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
        permissions:
          - actions:
              - "Microsoft.Compute/virtualMachines/read"
            data_actions:
              - "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write"
        assignable_scopes:
            - "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
      register: output

    - name: Assert creating role definition
      assert:
        that:
          - output.changed
    - name: Get facts
      azure_rm_roledefinition_facts:
        role_name: "{{ role_name }}"
        scope: "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
        type: "custom"
      register: facts
    - name: dump output
      debug:
        var: facts

   - name: lsit
      azure_rm_roledefinition_facts:
        scope: "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
        type: custom
      register: output

    - name: dump output
      debug:
        var: output

    - name: get by name
      azure_rm_roledefinition_facts:
        scope: "/subscriptions/{{ subscription_id }}/resourceGroups/{{ resource_group }}"
        role_name: "{{ role_name }}"
      register: output

    - name: dump output
      debug:
        var: output
```

facts output snippet
```
TASK [Create a role definition (Check Mode)] **************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:19
Tuesday 26 February 2019  13:35:11 +0800 (0:00:02.344)       0:00:04.821 ******
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308 `" && echo ansible-tmp-1551159311.9925406-89353712743308="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-1017475mac0wz/tmpwcnl7k91 TO /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308/ /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551159311.9925406-89353712743308/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Azure API profile latest does not define an entry for AuthorizationManagementClient

changed: [localhost] => changed=true
  id: null
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: xxxxxxxx-007d-457d-abb4-c31803a8857e
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
        not_actions: null
        not_data_actions: null
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [Assert creating role definition check mode] *********************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:33
Tuesday 26 February 2019  13:35:14 +0800 (0:00:02.954)       0:00:07.775 ******
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Create a role definition] ***************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:38
Tuesday 26 February 2019  13:35:14 +0800 (0:00:00.055)       0:00:07.831 ******
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418 `" && echo ansible-tmp-1551159315.003723-17541038200418="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-1017475mac0wz/tmpfdlljsez TO /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418/ /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551159315.003723-17541038200418/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: xxxxxxxx-007d-457d-abb4-c31803a8857e
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
        not_actions: null
        not_data_actions: null
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [Assert creating role definition] ********************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:51
Tuesday 26 February 2019  13:35:25 +0800 (0:00:10.638)       0:00:18.470 ******
ok: [localhost] => changed=false
  msg: All assertions passed


TASK [lsit] ***********************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:19
Tuesday 26 February 2019  13:29:42 +0800 (0:00:03.178)       0:00:04.986 ******
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161 `" && echo ansible-tmp-1551158982.6961415-99415513096161="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition_facts.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-9958zt5ltdlz/tmpxf795vpq TO /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161/AnsiballZ_azure_rm_roledefinition_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161/ /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551158982.6961415-99415513096161/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Azure API profile latest does not define an entry for AuthorizationManagementClient

ok: [localhost] => changed=false
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      id: null
      password: null
      profile: null
      role_name: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      subscription_id: null
      tenant: null
      type: custom
  roledefinitions:
  - assignable_scopes:
    - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
    description: null
    id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-007d-457d-abb4-c31803a8857e
    name: xxxxxxxx-007d-457d-abb4-c31803a8857e
    permissions:
    - actions:
      - Microsoft.Compute/virtualMachines/read
      data_actions:
      - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      not_actions: []
      not_data_actions: []
    role_name: xxxxxxxx-ec64-4c98-8fb0-54dd09ab2a99
    type: CustomRole

TASK [dump output] ****************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:25
Tuesday 26 February 2019  13:29:45 +0800 (0:00:02.421)       0:00:07.408 ******
ok: [localhost] =>
  output:
    changed: false
    failed: false
    roledefinitions:
    - assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      description: null
      id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-007d-457d-abb4-c31803a8857e
      name: xxxxxxx-007d-457d-abb4-c31803a8857e
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
        not_actions: []
        not_data_actions: []
      role_name: xxxxxxx-ec64-4c98-8fb0-54dd09ab2a99
      type: CustomRole
    warnings:
    - Azure API profile latest does not define an entry for AuthorizationManagementClient

TASK [get by name] ****************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:29
Tuesday 26 February 2019  13:29:45 +0800 (0:00:00.057)       0:00:07.465 ******
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079 `" && echo ansible-tmp-1551158985.1766756-93796864095079="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition_facts.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-9958zt5ltdlz/tmp1j3ld3wr TO /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079/AnsiballZ_azure_rm_roledefinition_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079/ /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551158985.1766756-93796864095079/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => changed=false
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      id: null
      password: null
      profile: null
      role_name: xxxxxxxxx-ec64-4c98-8fb0-54dd09ab2a99
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      subscription_id: null
      tenant: null
      type: null
  roledefinitions:
    assignable_scopes:
    - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
    description: null
    id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxx-007d-457d-abb4-c31803a8857e
    name: xxxxxxx-007d-457d-abb4-c31803a8857e
    permissions:
    - actions:
      - Microsoft.Compute/virtualMachines/read
      data_actions:
      - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      not_actions: []
      not_data_actions: []
    role_name: xxxxxxx-ec64-4c98-8fb0-54dd09ab2a99
    type: CustomRole

TASK [dump output] ****************************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:35
Tuesday 26 February 2019  13:29:47 +0800 (0:00:02.780)       0:00:10.246 ******
ok: [localhost] =>
  output:
    changed: false
    failed: false
    roledefinitions:
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      description: null
      id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxx-007d-457d-abb4-c31803a8857e
      name: xxxxxxx-007d-457d-abb4-c31803a8857e
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
        not_actions: []
        not_data_actions: []
      role_name: xxxxxxxx-ec64-4c98-8fb0-54dd09ab2a99
      type: CustomRole
    warnings:
    - Azure API profile latest does not define an entry for AuthorizationManagementClient
```

yungezz commented:
ready_for_review

yungezz committed:
fix lint

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition.py:
```diff
@@ -0,0 +1,401 @@
+                                                base_url=self._cloud_environment.endpoints.resource_manager,
+                                                api_version="2018-01-01-preview")
+
+        # build cope
```
scope.... 
maybe this comment is redundant anyway ;-)

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition.py:
```diff
@@ -0,0 +1,401 @@
+
+                self.log('role definition deleted')
+
+            else:
```
why fail here? I think we should just do nothing

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition_facts.py:
```diff
@@ -0,0 +1,298 @@
+EXAMPLES = '''
+    - name: List Role Definitions in scope
+      azure_rm_roledefinition_facts:
+        scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/myresourceGroup
```
just for consistency: myresourceGroup -> myResourceGroup

zikalino commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition_facts.py:
```diff
@@ -0,0 +1,298 @@
+            data_actions:
+                description:
+                    - List of allowed data actions.
+                returned: always
```
maybe add "type: list"

zikalino reviewed (changes requested):
nothing major, just that fail when delete should be fixed i think.

yungezz committed:
resolve comments

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition.py:
```diff
@@ -0,0 +1,400 @@
+    def permissions_to_set(self, permissions):
+        new_permissions = [str(dict(
+            actions=(set([to_native(a) for a in item.get('actions')]) if item.get('actions') else None),
+            non_actions=(set([to_native(a) for a in item.get('non_actions')]) if item.get('non_actions') else None),
```
Shouldn't these be `not_actions` ? ... and the tests should probably include some of these to catch the fact that they're not working... ;)

nitzmahone commented on lib/ansible/modules/cloud/azure/azure_rm_roledefinition.py:
```diff
@@ -0,0 +1,400 @@
+            actions=(set([to_native(a) for a in item.get('actions')]) if item.get('actions') else None),
+            non_actions=(set([to_native(a) for a in item.get('non_actions')]) if item.get('non_actions') else None),
+            data_actions=(set([to_native(a) for a in item.get('data_actions')]) if item.get('data_actions') else None),
+            non_data_actions=(set([to_native(a) for a in item.get('non_data_actions')]) if item.get('non_data_actions') else None),
```
`not_data_actions` + tests (same as above)

nitzmahone reviewed (changes requested):
looks like not_*actions are broken and not tested; once that's fixed, LGTM

yungezz commented:
test result after resolving comments:

```
TASK [Create a role definition (Check Mode)] *************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:19
Thursday 28 February 2019  11:26:17 +0800 (0:00:02.285)       0:00:03.780 *****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670 `" && echo ansible-tmp-1551324377.7568047-75761710646670="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-54514icrrtn_/tmp1urfiy0f TO /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670/ /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551324377.7568047-75761710646670/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Azure API profile latest does not define an entry for AuthorizationManagementClient

changed: [localhost] => changed=true
  id: null
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: yungez0228role
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
        not_actions:
        - Microsoft.Compute/virtualMachines/write
        not_data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [Assert creating role definition check mode] ********************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:37
Thursday 28 February 2019  11:26:20 +0800 (0:00:02.979)       0:00:06.760 *****
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Create a role definition] **************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:42
Thursday 28 February 2019  11:26:20 +0800 (0:00:00.055)       0:00:06.816 *****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687 `" && echo ansible-tmp-1551324380.7901793-185390353818687="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-54514icrrtn_/tmpfjnzb2qd TO /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687/ /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551324380.7901793-185390353818687/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/2e6f64dc-xxxx-xxxx-xxxx-d68365d38a40
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: yungez0228role
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
        not_actions:
        - Microsoft.Compute/virtualMachines/write
        not_data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [Assert creating role definition] *******************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:60
Thursday 28 February 2019  11:26:27 +0800 (0:00:07.174)       0:00:13.991 *****
ok: [localhost] => changed=false
  msg: All assertions passed

  TASK [Get facts] *****************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:64
Thursday 28 February 2019  11:30:16 +0800 (0:00:03.004)       0:00:04.516 *****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659 `" && echo ansible-tmp-1551324616.7636352-27074515237659="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition_facts.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-6070d65d5ft5/tmpba1k0lqa TO /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659/AnsiballZ_azure_rm_roledefinition_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659/ /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659/AnsiballZ_azure_rm_roledefinition_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551324616.7636352-27074515237659/ > /dev/null 2>&1 && sleep 0'
 [WARNING]: Azure API profile latest does not define an entry for AuthorizationManagementClient

ok: [localhost] => changed=false
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      id: null
      password: null
      profile: null
      role_name: yungez0228role
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      subscription_id: null
      tenant: null
      type: custom
  roledefinitions:
    assignable_scopes:
    - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
    description: null
    id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-d68365d38a40
    name: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
    permissions:
    - actions:
      - Microsoft.Compute/virtualMachines/read
      data_actions:
      - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
      not_actions:
      - Microsoft.Compute/virtualMachines/write
      not_data_actions:
      - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
    role_name: yungez0228role
    type: CustomRole

TASK [dump output] ***************************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:71
Thursday 28 February 2019  11:30:19 +0800 (0:00:03.179)       0:00:07.696 *****
ok: [localhost] =>
  facts:
    changed: false
    failed: false
    roledefinitions:
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      description: null
      id: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/providers/Microsoft.Authorization/roleDefinitions/xxxxxxxx-xxxx-xxxx-xxxx-d68365d38a40
      name: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
        not_actions:
        - Microsoft.Compute/virtualMachines/write
        not_data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      role_name: yungez0228role
      type: CustomRole
    warnings:
    - Azure API profile latest does not define an entry for AuthorizationManagementClient

TASK [Update the role definition (idempotent)] ***********************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:75
Thursday 28 February 2019  11:30:19 +0800 (0:00:00.065)       0:00:07.762 *****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190 `" && echo ansible-tmp-1551324620.009144-14330818308190="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-6070d65d5ft5/tmpirs8mmw0 TO /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190/ /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551324620.009144-14330818308190/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => changed=false
  id: /subscriptions/4933a9cb-6207-4501-96e3-23a04a383ad0/providers/Microsoft.Authorization/roleDefinitions/2e6f64dc-461d-45d4-9027-d68365d38a40
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: yungez0228role
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
        not_actions:
        - Microsoft.Compute/virtualMachines/write
        not_data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [assert output not changed] *************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:93
Thursday 28 February 2019  11:30:22 +0800 (0:00:02.891)       0:00:10.654 *****
ok: [localhost] => changed=false
  msg: All assertions passed

TASK [Update the role definition] ************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:98
Thursday 28 February 2019  11:30:22 +0800 (0:00:00.058)       0:00:10.713 *****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: yungez
<127.0.0.1> EXEC /bin/sh -c 'echo ~yungez && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015 `" && echo ansible-tmp-1551324622.9607058-158560230777015="` echo /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015 `" ) && sleep 0'
Using module file /home/yungez/.ansible/roles/azure_preview_modules/library/azure_rm_roledefinition.py
<127.0.0.1> PUT /home/yungez/.ansible/tmp/ansible-local-6070d65d5ft5/tmp64_aj6pi TO /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015/AnsiballZ_azure_rm_roledefinition.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015/ /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/home/yungez/src/ansible/venv/bin/python /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015/AnsiballZ_azure_rm_roledefinition.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/yungez/.ansible/tmp/ansible-tmp-1551324622.9607058-158560230777015/ > /dev/null 2>&1 && sleep 0'
changed: [localhost] => changed=true
  id: /subscriptions/4933a9cb-6207-4501-96e3-23a04a383ad0/providers/Microsoft.Authorization/roleDefinitions/e6abde0a-6126-44f6-8bbe-d769847cf8c1
  invocation:
    module_args:
      ad_user: null
      adfs_authority_url: null
      api_profile: latest
      assignable_scopes:
      - /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      auth_source: null
      cert_validation_mode: null
      client_id: null
      cloud_environment: AzureCloud
      description: null
      name: yungez0228role1
      password: null
      permissions:
      - actions:
        - Microsoft.Compute/virtualMachines/read
        - Microsoft.Compute/virtualMachines/start/action
        data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read
        not_actions:
        - Microsoft.Compute/virtualMachines/write
        not_data_actions:
        - Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write
      profile: null
      scope: /subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx/resourceGroups/ansible-roledefinition1
      secret: null
      state: present
      subscription_id: null
      tenant: null

TASK [assert output changed] *****************************************************************************************************************************************************************
task path: /home/yungez/src/ansible-testapp/roledefinition.yml:117
Thursday 28 February 2019  11:30:34 +0800 (0:00:11.249)       0:00:21.962 *****
ok: [localhost] => changed=false
  msg: All assertions passed


```

yungezz committed:
fix not_xxx_actions

yungezz commented:
ready_for_review

Fred-sun commented:
@nitzmahone Please help review this PR when you're available! Thanks!
